src/TableFunctions/TableFunctionS3Distributed.cpp

```diff
@@ -0,0 +1,143 @@
+#include <thread>
+#include <Common/config.h>
+#include "DataStreams/RemoteBlockInputStream.h"
+#include "Interpreters/ClientInfo.h"
+#include "Parsers/ASTExpressionList.h"
+#include "Parsers/ASTFunction.h"
+#include "Parsers/IAST_fwd.h"
+#include "Processors/Sources/SourceFromInputStream.h"
+#include "Storages/StorageS3Distributed.h"
+
+#if USE_AWS_S3
+
+#include <IO/S3Common.h>
+#include <Storages/StorageS3.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Interpreters/Context.h>
+#include <TableFunctions/TableFunctionFactory.h>
+#include <TableFunctions/TableFunctionS3.h>
+#include <TableFunctions/TableFunctionS3Distributed.h>
+#include <TableFunctions/parseColumnsListForTableFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include "registerTableFunctions.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+void TableFunctionS3Distributed::parseArguments(const ASTPtr & ast_function, const Context & context)
+{
+    /// Parse args
+    ASTs & args_func = ast_function->children;
+
+    if (args_func.size() != 1)
+        throw Exception("Table function '" + getName() + "' must have arguments.", ErrorCodes::LOGICAL_ERROR);
+
+    ASTs & args = args_func.at(0)->children;
+
+    if (args.size() < 4 || args.size() > 7)
+        throw Exception("Table function '" + getName() + "' requires 4 to 7 arguments: cluster, url," + 
+            "[access_key_id, secret_access_key,] format, structure and [compression_method].",
+            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    for (auto & arg : args)
+        arg = evaluateConstantExpressionOrIdentifierAsLiteral(arg, context);
+
+    cluster_name = args[0]->as<ASTLiteral &>().value.safeGet<String>();
+    filename_or_initiator_hash = args[1]->as<ASTLiteral &>().value.safeGet<String>();
+
+    if (args.size() < 5)
+    {
+        format = args[2]->as<ASTLiteral &>().value.safeGet<String>();
+        structure = args[3]->as<ASTLiteral &>().value.safeGet<String>();
+    }
+    else
+    {
+        access_key_id = args[2]->as<ASTLiteral &>().value.safeGet<String>();
+        secret_access_key = args[3]->as<ASTLiteral &>().value.safeGet<String>();
+        format = args[4]->as<ASTLiteral &>().value.safeGet<String>();
+        structure = args[5]->as<ASTLiteral &>().value.safeGet<String>();
```

Comment:
If `args.size() == 5`, it will segfault

Reply:
Rewrited this to a bit bigger but simplier code